### DISCUSSION TITLE
How to enable SPDY support

### DISCUSSION THREAD

**Discussion by landorg** [Kong/kong#7334] [Ideas and feature requests]
### Summary

According to [this](http://nginx.org/en/docs/http/ngx_http_spdy_module.html) there is some kind of SPDY support in nginx now. How can I enable it in kong? 
Could not find any documentation about it and the [community attempt](https://discuss.konghq.com/t/spdy-support-for-kubectl/7474) failed too.
I'm trying to forward Rancher traffic with kong but kubectl exec or port-forward are not using when going through the "rancher kube-api".

### Additional Details & Logs

- Kong version (`2.1`)

**Comment by kikito** [comment]
Hello, currently there is no direct support for SPDY in Kong.

You might be able to route its traffic using the [TCP stream support](https://konghq.com/blog/tcp-stream-support-kong/).

**Comment by dndx** [comment]
Hello @rolandg,

The official Kong release does not include the [ngx_http_spdy_module](http://nginx.org/en/docs/http/ngx_http_spdy_module.html) module which is required for SPDY support. If you want to use SPDY, my best guess is to [build Kong yourself](https://docs.konghq.com/install/source/) with the module enabled and modify the `nginx.conf` template of Kong to enable it.

If you only need to proxy SPDY traffic without needing to inspect the content, then what @kikito proposed is likely easier and will work with standard Kong builds.

**Comment by landorg** [comment]
Thanks for your answers. Going to try out the tcp stream in a bit.

**Comment by landorg** [comment]
I found out the hard way what
> This module was superseded by the ngx_http_v2_module module in 1.9.5.
means. Since it's also missing in the kubernetes ingress controller I'm wondering how are they doing.

Compared their nginx.conf templates a bit with what rancher recommends for plain nginx and it's not too far off.
I added what I thaught was missing (
```
map $upstream_connection $upstream_upgrade {
    default Upgrade;
    ''      close;
}
```
but without success.
Ideas anyone?

**Comment by minhnguyenvan95** [comment]
same issue when using kubectl through kong

direct to rancher response header
```
I0310 03:05:11.183367  248005 round_trippers.go:577] Response Headers:
I0310 03:05:11.183375  248005 round_trippers.go:580]     Content-Length: 0
I0310 03:05:11.183392  248005 round_trippers.go:580]     Connection: Upgrade
I0310 03:05:11.183414  248005 round_trippers.go:580]     Date: Thu, 10 Mar 2022 08:05:11 GMT
I0310 03:05:11.183433  248005 round_trippers.go:580]     Upgrade: SPDY/3.1
I0310 03:05:11.183454  248005 round_trippers.go:580]     X-Api-Cattle-Auth: true
I0310 03:05:11.183472  248005 round_trippers.go:580]     X-Content-Type-Options: nosniff
I0310 03:05:11.183490  248005 round_trippers.go:580]     X-Stream-Protocol-Version: portforward.k8s.io
```

through kong response header
```
I0310 15:05:49.033798   85687 round_trippers.go:460] Response Headers:
I0310 15:05:49.033810   85687 round_trippers.go:463]     Date: Thu, 10 Mar 2022 08:05:49 GMT
I0310 15:05:49.033817   85687 round_trippers.go:463]     Via: kong/2.7.1
I0310 15:05:49.033823   85687 round_trippers.go:463]     Content-Type: application/json
I0310 15:05:49.033829   85687 round_trippers.go:463]     Content-Length: 139
I0310 15:05:49.033834   85687 round_trippers.go:463]     Cache-Control: no-cache, private
I0310 15:05:49.033841   85687 round_trippers.go:463]     X-Api-Cattle-Auth: true
I0310 15:05:49.033849   85687 round_trippers.go:463]     X-Content-Type-Options: nosniff
I0310 15:05:49.033855   85687 round_trippers.go:463]     X-Kong-Upstream-Latency: 10
I0310 15:05:49.033860   85687 round_trippers.go:463]     X-Kong-Proxy-Latency: 1
I0310 15:05:49.033866   85687 round_trippers.go:463]     Connection: keep-alive
I0310 15:05:49.033872   85687 round_trippers.go:463]     Audit-Id: 51b633e2-80c4-4d44-9dfd-83a9a5f73e27
F0310 15:05:49.034503   85687 helpers.go:115] error: error upgrading connection: Upgrade request required
goroutine 1 [running]:
```

Related https://www.spectrocloud.com/blog/kubectl-exec-through-https-tunnel-and-reverse-proxy/
